### DISCUSSION TITLE
including CO2 emissions in a multi parameter optimization

### DISCUSSION THREAD

**Discussion by ayp234** [FZJ-IEK3-VSA/FINE#55] [Q&A]
I would like to examine the trade-offs in a multi parameter optimization of a company's energy supply system. So far, FINE can optimize OPEX+CAPEX, and set constraints on emissions (i. e. model the commodity CO2 as a sink with an annual limit). Are there plans to extend the objective function to include emissions? In the meantime, do you have an idea for a workaround?

- [ ] write a heuristic optimization algorithm that tries out different constraints and uses the output of the FINE model (TAC) as input for a new objective function which consists of TAC and GHG emissions (weighted and normalised by the base case scenario)
- [ ] monetize CO2, so it becomes part of OPEX
- [ ] use pyomo to directly calculate a new objective function instead of using FINE's energySystemModel.optimize.....

**Comment by t-gross** [comment]
There are plans to extend the objective functions but it will probably require some time to finish this. 
Your ideas are already looking quite good. Your second idea will probably be the easiest to implement.
For the third one: 
The pyomo instance can directly be overwritten. It´s stored in energySystemModel.pyM after you run the function _declareOptimizationProblem_ on the energySystemModel instance.

**Reply by ayp234** [reply]
I followed this workflow:
-  Required packages are imported
-  An energy system model instance is created
-  Commodity sources are added to the energy supply system model
-  Commodity conversion components are added to the energy supply system model
-  Commodity storages are added to the energy supply system model
-  Commodity transmission components are added to the energy supply system model
-  Commodity sinks are added to the energy supply system model, **including a sink called "CO2 to environment"**

According to your response, I should then take the following steps:

esM.declareOptimizationProblem()
pyM = esM.pyM
def new_objective(pyM):
    NPV = sum(
        mdl.getObjectiveFunctionContribution(esM, pyM)
        for mdl in esM.componentModelingDict.values()
    )

    co2_emissions = sum(
       **How can I access the operation variable of the Sink Class "CO2 to environment"?** I am referring to the operation variable which is also printed if one calls getOptimizationSummary. I couldn't trace it back in the source code.
    )

    return NPV + co2_emissions _(still needs to be normalised somehow)_

pyM.Obj = pyomo.Objective(rule=new_objective)
esM.optimize(declareOptimizationProblem=False)
results = esM.getOptimizationSummary('SourceSinkModel', outputLevel=2)

I also thought about not changing the source code but adding directly to esM.pyM.Obj. I thought about somehow adding the yearlyLimit of the Sink "CO2 to environment" to the Objective and Variables.
